firehose.go

```diff
@@ -292,10 +309,61 @@ func (fp *Firehose) moveFilesToOutputBucket(ctx context.Context, filePathPrefix
 			// ignore errors when deleting, picked up again on next run
 			_ = fp.StorageBackend.DeleteFile(ctx, fp.processSettings.StagingBucket, fileName)
 		}
+
+		err = fp.writeIndex(ctx, filePathPrefix, currTime, filename, minTimestamp, maxTimestamp)
+		if err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
+// maintain an index file (one for every output directory) listing all the data files and their timestamp range
+func (fp *Firehose) writeIndex(ctx context.Context, filePathPrefix string, currTime time.Time, name string, minTimestamp int64, maxTimestamp int64) error {
+	indexEntry, err := json.Marshal(indexFile{
+		Name:         name,
+		MinTimestamp: minTimestamp,
+		MaxTimestamp: maxTimestamp,
+	})
+	indexEntry = append(indexEntry, '\n')
+	if err != nil {
+		return err
+	}
+	indexLocation := fmt.Sprintf("%s/%s/_metadata.ndjson", filePathPrefix, currTime.Format("2006/1/2"))
+	// GCS objects are immutable: we cannot append to the index. So we read the old index,
```

Comment:
hmmm we may have to just try this wasn't able to get write down to 5 minutes so hopefully this file doesn't get too big

Reply:
Brandon pointed out that there's a more complicated scheme that we can use if the index gets big: after some threshold, we just start a new empty index file. The new index file contains a line pointing to the GCS generation of the previous index file, and so then on reading we can "follow the links" and concatenate the files together.

The nice thing about the more complicated scheme is that it's forwards-compatible with data written under the simple scheme. So we can see if it's a problem first and then do the more complicated thing if it is.

Reply:
ah I see makes sense yeah we also could just number them